// @aspectjs/core/dist/esm2015/commons/public_api.js
export { AnnotationType as AdviceType, AnnotationRef, AnnotationType } from './src/annotation/annotation.types.js';
export { AspectError } from './src/weaver/errors/aspect-error.js';
export { WeavingError } from './src/weaver/errors/weaving-error.js';
export { AdviceError } from './src/weaver/errors/advice-error.js';
export { AnnotationPointcutExpressionBuilder, Pointcut, PointcutExpression, PointcutPhase, PropertyAnnotationPointcutExpressionBuilder, on } from './src/types/pointcut.js';
export { _JoinpointFactory } from './src/weaver/joinpoint-factory.js';
export { WeaverProfile } from './src/weaver/profile.js';
export { _getWeaverContext, _setWeaverContext } from './src/weaver/weaver-context.js';
export { _AdviceFactory } from './src/advices/advice.factory.js';
export { AnnotationContext } from './src/annotation/context/annotation.context.js';
export { AnnotationRegistry } from './src/annotation/registry/annotation.registry.js';
export { AnnotationTargetFactory } from './src/annotation/target/annotation-target.factory.js';
export { AnnotationLocationFactory } from './src/annotation/location/location.factory.js';
export { ClassAnnotationsBundle, RootAnnotationsBundle } from './src/annotation/bundle/bundle.js';
export { AnnotationFactory } from './src/annotation/factory/annotation.factory.js';
export { ASPECTJS_ANNOTATION_FACTORY } from './src/aspectjs.annotation.factory.js';
